language/sygus.tex

```diff
@@ -2632,6 +2730,90 @@ \section{Examples}%
 and hence the solution for $f$ satisfies all input constraints.
 \end{example}
 
+
+\begin{example}[Optimization with Weights]
+Consider the following example:
+\begin{lstlisting}[language=SyGuS]
+(set-logic LIA)
+(set-feature :weights true)
+(declare-weight branches)
+(synth-fun f ((x Int)) Int
+   ((I Int) (B Bool))
+   ((I Int (0 1 x (+ I I) (! (ite B I I) :branches 1)))
+    (B Bool ((>= I I) (= I I)))))
+(declare-var x Int)
+(constraint (= (f x) (ite (= x 0) 0 x)))
+(optimize-synth ((! (_ branches f) :min)))
+\end{lstlisting}
+In this example, a weight keyword $\texttt{branches}$ has been declared.
+The grammar of function $\texttt{f}$ has been annotated such that
+$\texttt{branches}$ corresponds to the number of $\texttt{ite}$ terms in its body.
+An optimization query is made, where
+the objective is to minimize the number of branches in $\texttt{f}$.
+This is specified by a term annotation on $\texttt{(\_ branches f)}$.
+A possible correct response for this example is:
+\begin{lstlisting}[language=SyGuS]
+(
+(1)
+(define-fun f ((x Int)) Int (ite (= x 0) 0 x))
+)
+\end{lstlisting}
+Another possible response for this example is:
+\begin{lstlisting}[language=SyGuS]
+(
+(0)
+(define-fun f ((x Int)) Int x)
+)
+\end{lstlisting}
+In both cases, the value of term $\texttt{(\_ branches f)}$ was given
+in the solution that was consistent with the number of
+$\texttt{ite}$ terms in the body of $\texttt{f}$.
+Hence, both responses are correct.
+The latter solution gives a smaller value for this term,
+and hence it is preferred based on the specified objective.
+
+Notice that in this example,
+no attribute is provided for the $\texttt{optimize-synth}$ command,
+meaning that no ordering is specified on the tuple passes as argument
+to this command.
+For details on the default semantics for objectives, see \cref{sec:objectives}.
+In the next example, we show how ordering on tuples can be specified.
+\end{example}
+
+\begin{example}[Lexicographic Optimization]
+Consider the following example:
+\begin{lstlisting}[language=SyGuS]
+(set-logic LIA)
+(synth-fun f ((x Int)) Int
+   ((I Int) (B Bool))
+   ((I Int (0 1 x (+ I I) (ite B I I)))))
+(declare-var x Int)
+(constraint (or (= (f x) 1) (= (f x) x)))
+(optimize-synth ((! (f 0) :max) (! (f 100) :max)) :lexicographic)
+\end{lstlisting}
+This example also contains an optimization query.
+The objective of the query is to optimize the values
+of terms $\texttt{(f 0)}$ and $\texttt{(f 100)}$.
+Moreover, the ordering on tuples of these terms is lexicographic.
+A possible correct response for this example is:
+\begin{lstlisting}[language=SyGuS]
+(
+(1 1)
+(define-fun f ((x Int)) Int 1)
+)
+\end{lstlisting}
+Another possible response for this example is:
+\begin{lstlisting}[language=SyGuS]
+(
+(0 100)
+(define-fun f ((x Int)) Int x)
+)
+\end{lstlisting}
+The first example is preferred, since the first value in the first
```

Comment:
I am unclear about this part. Why is `(1,1)` preferred over `(0,100)` lexicographically? Wouldn't `(0,100)` appear before `(1,1)` in lexicographic ordering?

I guess the `:lexicographic` indicates lexicographic `>` ordering. Is that correct?

Reply:
It is lexicographic based on the orderings for each term position. In this case, we have said that one should try to maximize the values for each term position, so (1,1) is preferred to (0,100).
Notice this would also extend to cases like:
`(optimize-synth ((! t :max) (! s :min)) :lexicographic)`

Reply:
Ah I understand now! Thank you for the explanation.